### ISSUE TITLE
question about how transform is implemented

### ISSUE THREAD

**Issue by DougBurke** [gicentre/elm-vegalite#27]
In adding in support for `density` and `loess` I was reminded that I wanted to ask you about how `transform` is implemented. At present you have

- routines like `density` that generate a JSON list of properties, where the ordering depends on the transform type
- `transform` takes this and generates the actual Vega Lite JSON object 

In looking at this, I don't see why `density` doesn't just create the Vega Lite JSON object directly. There is one advantage to the current approach, in that `transform` can skip any input it doesn't know about, but I think that could be added to my approach.

An example, instead of

```
density field dnProps =
    (::)
        ( "density"
        , toList
            [ JE.string field
            , densityPropertySpec "groupby" dnProps
            , densityPropertySpec "cumulative" dnProps
            , densityPropertySpec "counts" dnProps
            , densityPropertySpec "bandwidth" dnProps
            , densityPropertySpec "extent" dnProps
            , densityPropertySpec "minsteps" dnProps
            , densityPropertySpec "maxsteps" dnProps
            , densityPropertySpec "steps" dnProps
            , densityPropertySpec "as" dnProps
            ]
        )
```

have `density` create the object from the following code (but without the need to decode the JSON since we know what options were defined by the user):

```
                "density" ->
                    case JD.decodeString (JD.list JD.value) (JE.encode 0 val) of
                        Ok [ dn, gbObj, cmObj, cnObj, bwObj, exObj, mnsObj, mxsObj, sObj, asObj ] ->
                            ([ ( "density", dn ) ]
                                ++ (if gbObj == JE.null then
                                        []

                                    else
                                        [ ( "groupby", gbObj ) ]
                                   )
                                ++ (if cnObj == JE.null then
                                        []

                                    else
                                        [ ( "cumulative", cnObj ) ]
                                   )
                                ++ (if cnObj == JE.null then
                                        []

                                    else
                                        [ ( "counts", cnObj ) ]
                                   )
                                ++ (if bwObj == JE.null then
                                        []

                                    else
                                        [ ( "bandwidth", bwObj ) ]
                                   )
                                ++ (if exObj == JE.null then
                                        []

                                    else
                                        [ ( "extent", exObj ) ]
                                   )
                                ++ (if mnsObj == JE.null then
                                        []

                                    else
                                        [ ( "minsteps", mnsObj ) ]
                                   )
                                ++ (if mxsObj == JE.null then
                                        []

                                    else
                                        [ ( "maxSteps", mxsObj ) ]
                                   )
                                ++ (if sObj == JE.null then
                                        []

                                    else
                                        [ ( "steps", sObj ) ]
                                   )
                                ++ (if asObj == JE.null then
                                        []

                                    else
                                        [ ( "as", asObj ) ]
                                   )
                            )
                                |> JE.object

                        _ ->
                            JE.null
```

**Comment by jwoLondon** [comment]
Thanks for pointing this out. I am struggling to recall my original rationale for creating these internal lists and decoding them. It may have been my early clumsy attempt to collapse multiple specifications into a single labelled spec, but I agree it appears to involve unnecessary encoding/decoding steps.
 
Because each transform function (`density`, `loess` etc.) involves adding a single `labelledSpec` to the ordered list of transforms (and I don't want to introduce a breaking change by changing the type sigs), we still need a way of grouping a collection of labelled specs, so I propose we create an internal labelledSpec in the form `("multSpecs", obj)` where `obj` is the object storing the transform's parameters. As you say, this can then be generated in the transform function directly without relying on a list with fragile parameter orderings and null parameters to preserve sequence.

The pattern would then be something like this for any transform that returns more than one labelled specification (using the `aggregate` transform as an example):

```elm
aggregate : List Spec -> List String -> List LabelledSpec -> List LabelledSpec
aggregate ops groups =
    (::)
        ( "multiSpecs"
        , JE.object [ ( "aggregate", toList ops ), ( "groupby", JE.list JE.string groups )]
        )
```

The `transform` function would then be much simplified in that it just has to distinguish single from multiple spec transforms:

```elm
transform : List LabelledSpec -> ( VLProperty, Spec )
transform transforms =
     let
        assemble : LabelledSpec -> Spec
        assemble ( trName, val ) =
            case trName of
                "multiSpecs" ->
                    val
                 _ ->
                     JE.object [ ( trName, val ) ]
    in
    ( VLTransform, JE.list assemble transforms )
```

Does that seem sensible and would it work for you?

**Comment by DougBurke** [comment]
As I'm less concerned with API stability (due to having a smaller user base, and because I want to see how it works out), I went for an approach that creates a new type (`TransformSpec`) and then something like the following pseudo-Elm code

```
type TransformSpec = Spec

transform :: List TransformSpec -> ( VLProperty, Spec )
transform [] = ( VLTransform, JE.null )
transform ts = ( VLTransform, spec )

aggregate : List Spec -> List String -> List TransformSpec -> List TransformSpec
aggregate ops groups ols =
  (::) ( JE.object [ ("aggregate", toList ops)
                         , ("groupby", JE.list JE.string groups) ] )
```

For me I now have a set of functions I know can only be used with `transform`, which I think helps a user. I've also extended this to other routines, such as `encoding`, `configure`, `resolve`, `selection`. These are simpler than the transform version since in this case they are really just aliases for `LabelledSpec` as we still need the key and the value.

As you don't want to break your API - and I'm not sure this fits in with your overall design or Elm - I think your approach looks fine (I might argue that all logic for creating the object is in the routines and that transform just assembles them - ie it just strips out the label/trName value of the input - but I'm not sure it really matters).

For reference, the approach I'm playing with can be seen in https://github.com/DougBurke/hvega/pull/108 but not sure it really helps here. I don't think that any of the approaches are that hard to convert between for me.
